### PR TITLE
ORCA-151: As a developer, the orca bucket should be configurable at the collection level to better organize data.

### PR DESCRIPTION
**Pull request by AndrewDorn**
## Summary of Changes

Addresses [ORCA-151: As a developer, the orca bucket should be configurable at the collection level to better organize data.](https://bugs.earthdata.nasa.gov/browse/ORCA-151)

## Changes

* Added default_orca_bucket_override to copy_to_glacier and request_files

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## How Has This Been Tested?

Tests pass locally.

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the necessary documentation (remove those that do not apply)
    - [x] CHANGELOG.md
    - [x] Testing documentation
    - [x] Development documentation
    - [x] User documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
    - [x] Unit tests
- [x] New and existing unit tests pass locally with my changes
    - [x] Automated tests passing (if not fork)
- [x] Any dependent changes have been merged and published in downstream modules
- [x] My code has passed security scanning
    - [x] Snyk
    - [x] git-secrets

### CODE REVIEW

File: modules/workflows/OrcaRecoveryWorkflow/orca_recover_workflow.asl.json
Comment (bhazuka) [inline modules/workflows/OrcaRecoveryWorkflow/orca_recover_workflow.asl.json:56]: Need to change `"orcaDefaultBucketOverride": "{$.meta.collection.meta.orca_default_bucket}"` to 
`"orcaDefaultBucketOverride": "{$.meta.collection.meta.orcaDefaultBucketOverride}"` to be consistent with our other items in collection configuration.

File: modules/workflows/OrcaCopyToGlacierWorkflow/orca_copy_to_glacier_workflow.asl.json
Comment (bhazuka) [inline modules/workflows/OrcaCopyToGlacierWorkflow/orca_copy_to_glacier_workflow.asl.json:12]: Need to change `"orcaDefaultBucketOverride": "{$.meta.collection.meta.orca_default_bucket}"` to 
`"orcaDefaultBucketOverride": "{$.meta.collection.meta.orcaDefaultBucketOverride}"` to be consistent with our other items in collection configuration.

File: modules/workflows/OrcaRecoveryWorkflow/orca_recover_workflow.asl.json
Comment (bhazuka) [inline modules/workflows/OrcaRecoveryWorkflow/orca_recover_workflow.asl.json:53]: Is `glacier-bucket` still needed or does this go away now?

File: README.md
Comment (bhazuka) [inline README.md:262]: Need to change `"orcaDefaultBucketOverride": "{$.meta.collection.meta.orca_default_bucket}"` to 
`"orcaDefaultBucketOverride": "{$.meta.collection.meta.orcaDefaultBucketOverride}"` to be consistent with our other items in collection configuration.

File: CHANGELOG.md
Comment (bhazuka) [inline CHANGELOG.md:30]: Need to change `"orcaDefaultBucketOverride": "{$.meta.collection.meta.orca_default_bucket}"` to 
`"orcaDefaultBucketOverride": "{$.meta.collection.meta.orcaDefaultBucketOverride}"` to be consistent with our other items in collection configuration.

File: tasks/copy_to_glacier/README.md
Comment (bhazuka) [inline tasks/copy_to_glacier/README.md:243]: Need to change `"orcaDefaultBucketOverride": "{$.meta.collection.meta.orca_default_bucket}"` to 
`"orcaDefaultBucketOverride": "{$.meta.collection.meta.orcaDefaultBucketOverride}"` to be consistent with our other items in collection configuration.

File: tasks/copy_to_glacier/copy_to_glacier.py
Comment (bhazuka) [inline tasks/copy_to_glacier/copy_to_glacier.py:79]: Not sure if this is needed. maybe simplify to `Name of the default ORCA S3 Glacier bucket which can be overridden  by configuration settings.` Thoughts?

File: tasks/copy_to_glacier/copy_to_glacier.py
Comment (bhazuka) [inline tasks/copy_to_glacier/copy_to_glacier.py:109]: Should create a card if we don't have one to do logging in this task instead of print statements.

Maybe state in message using default value? Like `f"{CONFIG_ORCA_DEFAULT_BUCKET_OVERRIDE_KEY} is not set. Using ORCA_DEFAULT_BUCKET value."`

File: tasks/copy_to_glacier/test/unit_tests/test_copy_to_glacier.py
Comment (bhazuka) [inline tasks/copy_to_glacier/test/unit_tests/test_copy_to_glacier.py:343]: This test works, just thinking out of the box a little bit, I wonder if it would be easier to throw the logic for setting the glacier bucket into a small function like `get_orca_archive_bucket(config)` that would return the proper bucket name. This would then simplify the test. That function then could be done throughout the other tasks with the same test essentially.

File: tasks/copy_to_glacier/test/unit_tests/test_copy_to_glacier.py
Comment (bhazuka) [inline tasks/copy_to_glacier/test/unit_tests/test_copy_to_glacier.py:380]: Another way to do this is to use the built in enumerate function like the following

```
for idx, source_bucket_name in enumerate(source_bucket_names):
    head_object_calls.append(
        call(Bucket=source_bucket_name, Key=source_keys[idx])
    )
...
```

File: tasks/request_files/request_files.py
Comment (bhazuka) [inline tasks/request_files/request_files.py:157]: Wondering if this should be an info rather than a warn. Thoughts? 

Maybe expand the message to `f"{CONFIG_ORCA_DEFAULT_BUCKET_OVERRIDE_KEY} is not set. Using {OS_ENVIRON_ORCA_DEFAULT_GLACIER_BUCKET_KEY} environment variable value."` or something like that anyway stating we are using the default value.

File: tasks/request_files/request_files.py
Comment (bhazuka) [inline tasks/request_files/request_files.py:154]: Would it make sense to pull this logic out into a small function for easier testing/replication? Then setting a variable that gets passed throughout?

File: tasks/request_files/test/unit_tests/test_request_files.py
Comment (bhazuka) [inline tasks/request_files/test/unit_tests/test_request_files.py:111]: May want a test to validate that default is taken if override not set.

File: tasks/copy_to_glacier/test/unit_tests/test_copy_to_glacier.py
Comment (bhazuka) [inline tasks/copy_to_glacier/test/unit_tests/test_copy_to_glacier.py:334]: May want a test to validate that default is taken if override not set as well.

File: website/docs/developer/deployment-guide/deployment-with-cumulus.md
Comment (bhazuka) [inline website/docs/developer/deployment-guide/deployment-with-cumulus.md:549]: Instead of `"glacier"` for the value maybe have this as a more realistic bucket name like `"prod_orca_worm"` or something like that anyway. I would really like for us to detach from the word `glacier` to something more generic since in AWS there are essentially 3 glacier storage classes now and if we ever need to go to Azure or Google cloud service storage it would be good to be somewhat generic in our approach so that we are not telegraphing the underneath to users.

**Review by bhazuka — CHANGES_REQUESTED** [review]
Mostly minor things throughout with naming to keep consistency. I did through a couple of ideas for maybe a minor rewrite that may make maintenance and testing easier but may require some rework in the code for you to look at.

File: website/docs/developer/deployment-guide/deployment-with-cumulus.md
Comment (AndrewDorn) [reply website/docs/developer/deployment-guide/deployment-with-cumulus.md:350]: Fixed

File: modules/workflows/OrcaRecoveryWorkflow/orca_recover_workflow.asl.json
Comment (AndrewDorn) [reply modules/workflows/OrcaRecoveryWorkflow/orca_recover_workflow.asl.json:56]: Fixed

File: modules/workflows/OrcaCopyToGlacierWorkflow/orca_copy_to_glacier_workflow.asl.json
Comment (AndrewDorn) [reply modules/workflows/OrcaCopyToGlacierWorkflow/orca_copy_to_glacier_workflow.asl.json:12]: Fixed

File: modules/workflows/OrcaRecoveryWorkflow/orca_recover_workflow.asl.json
Comment (AndrewDorn) [reply modules/workflows/OrcaRecoveryWorkflow/orca_recover_workflow.asl.json:53]: Unless we want to force a migration, it is still needed.

File: README.md
Comment (AndrewDorn) [reply README.md:262]: Fixed

File: CHANGELOG.md
Comment (AndrewDorn) [reply CHANGELOG.md:30]: Fixed

File: tasks/copy_to_glacier/README.md
Comment (AndrewDorn) [reply tasks/copy_to_glacier/README.md:243]: Fixed

File: tasks/copy_to_glacier/copy_to_glacier.py
Comment (AndrewDorn) [reply tasks/copy_to_glacier/copy_to_glacier.py:79]: That would actually be a longer description, and drops the reference to file name 'config'.

File: tasks/copy_to_glacier/copy_to_glacier.py
Comment (AndrewDorn) [reply tasks/copy_to_glacier/copy_to_glacier.py:109]: That work is in @rizbi-hassan 's PR, so we'll see how the order shakes out. Added extra line.

File: tasks/request_files/request_files.py
Comment (AndrewDorn) [reply tasks/request_files/request_files.py:157]: Eh, I think a warning is a stronger encourager to get workflows updated. Added line.

File: tasks/request_files/test/unit_tests/test_request_files.py
Comment (AndrewDorn) [reply tasks/request_files/test/unit_tests/test_request_files.py:111]: I treated the other tests as base cases, and this as testing the override.

File: tasks/copy_to_glacier/test/unit_tests/test_copy_to_glacier.py
Comment (AndrewDorn) [reply tasks/copy_to_glacier/test/unit_tests/test_copy_to_glacier.py:334]: I treated the other tests as base cases, and this as testing the override.

File: website/docs/developer/deployment-guide/deployment-with-cumulus.md
Comment (AndrewDorn) [reply website/docs/developer/deployment-guide/deployment-with-cumulus.md:549]: Sounds good. Updated.

File: tasks/request_files/request_files.py
Comment (bhazuka) [reply tasks/request_files/request_files.py]: ok, warning it is

**Review by bhazuka — COMMENTED** [review]
Looking good so far. A couple of outstanding comments to discuss.

File: tasks/copy_to_glacier/copy_to_glacier.py
Comment (bhazuka) [reply tasks/copy_to_glacier/copy_to_glacier.py:79]: I guess I didn't make the files config connection reading it. I am still not convinced that we need to state how or where it is is overridden here. That may be better served in the description above where we tell what it does and the tasks behavior. Thoughts?

File: modules/workflows/OrcaRecoveryWorkflow/orca_recover_workflow.asl.json
Comment (AndrewDorn) [reply modules/workflows/OrcaRecoveryWorkflow/orca_recover_workflow.asl.json:53]: Well, we'd be removing the reference to orca_default_bucket, right? So that could break somebody's setup.

File: tasks/copy_to_glacier/copy_to_glacier.py
Comment (AndrewDorn) [reply tasks/copy_to_glacier/copy_to_glacier.py:79]: I would prefer to keep the explanation of environment variables centralized.

File: tasks/copy_to_glacier/test/unit_tests/test_copy_to_glacier.py
Comment (AndrewDorn) [reply tasks/copy_to_glacier/test/unit_tests/test_copy_to_glacier.py:343]: Holding off due to potential merge conflicts.

File: tasks/copy_to_glacier/test/unit_tests/test_copy_to_glacier.py
Comment (AndrewDorn) [reply tasks/copy_to_glacier/test/unit_tests/test_copy_to_glacier.py:380]: Holding off due to potential merge conflicts.

File: tasks/copy_to_glacier/copy_to_glacier.py
Comment (AndrewDorn) [reply tasks/copy_to_glacier/copy_to_glacier.py:79]: Also, unsure what above description you are referring to.

File: tasks/request_files/request_files.py
Comment (AndrewDorn) [reply tasks/request_files/request_files.py]: Done

Comment (AndrewDorn) [reply tasks/request_files/request_files.py]: Done

File: tasks/copy_to_glacier/README.md
Comment (rizbi-hassan) [inline tasks/copy_to_glacier/README.md:396]: any reason why this pydoc for sqs_library is removed?

File: tasks/copy_to_glacier/copy_to_glacier.py
Comment (rizbi-hassan) [inline tasks/copy_to_glacier/copy_to_glacier.py:109]: I am guesing this will need to be changed to logger.info later once orca-230 is merged.

File: tasks/copy_to_glacier/copy_to_glacier.py
Comment (rizbi-hassan) [inline tasks/copy_to_glacier/copy_to_glacier.py:118]: I am guesing this will need to be changed to logger.info later once orca-230 is merged.

**Review by rizbi-hassan — COMMENTED** [review]
I would suggest we first try to merge orca-230 before merging these changes to avoid merge conflicts.

File: tasks/copy_to_glacier/README.md
Comment (AndrewDorn) [reply tasks/copy_to_glacier/README.md:396]: Whoops. Fixed.

File: modules/workflows/OrcaRecoveryWorkflow/orca_recover_workflow.asl.json
Comment (bhazuka) [reply modules/workflows/OrcaRecoveryWorkflow/orca_recover_workflow.asl.json:53]: I don't believe so, I think that may have been a placeholder. I would expect we would get the ORCA bucket name either from either the environment `ORCA_DEFAULT_BUCKET`, or the configuration `orcaDefaultBucketOverride` this third variable I feel is confusing and does not provide any benefit. Looking at the code for request_files it really isn't used except to check for it and set the value to our environment variable. Seeing that this is a part of our workflow that we distribute, I feel like we can get rid of it. Thoughts?

File: tasks/copy_to_glacier/copy_to_glacier.py
Comment (bhazuka) [reply tasks/copy_to_glacier/copy_to_glacier.py:79]: Referring to the overall definition starting on line 74. So changing `Copies the files in {event}['input']
    to the ORCA glacier bucket defined in ORCA_DEFAULT_BUCKET.` to `Copies the files in {event}['input']
    to the ORCA glacier bucket defined in ORCA_DEFAULT_BUCKET overridden by setting the orcaDefaultBucketOverride variable in the configuration at the collection or step function level.`

File: tasks/copy_to_glacier/test/unit_tests/test_copy_to_glacier.py
Comment (bhazuka) [reply tasks/copy_to_glacier/test/unit_tests/test_copy_to_glacier.py:380]: Let's make a card or have a plan if we want to implement this to help with the merge conflicts so we don't forget about it.

File: tasks/request_files/test/unit_tests/test_request_files.py
Comment (bhazuka) [inline tasks/request_files/test/unit_tests/test_request_files.py:1796]: Not sure why we are using Mock() here for bucket when this in reality is a string not an object. This should really just be a known string throughout these tests I would think. Thoughts?

**Review by bhazuka — CHANGES_REQUESTED** [review]
Couple of questions and comments

File: tasks/request_files/test/unit_tests/test_request_files.py
Comment (AndrewDorn) [reply tasks/request_files/test/unit_tests/test_request_files.py:1796]: I like mocks as a sort of added check that no processing is done with them. There shouldn't be any processing here, so this raises an error if, say, the order of checks is wrong.

Comment (AndrewDorn) [reply tasks/request_files/test/unit_tests/test_request_files.py:1796]: Side-pedantic-note, strings in Python are objects. :]

Comment (bhazuka) [reply tasks/request_files/test/unit_tests/test_request_files.py:1796]: Good point

File: tasks/copy_to_glacier/copy_to_glacier.py
Comment (AndrewDorn) [reply tasks/copy_to_glacier/copy_to_glacier.py:118]: Went with warning, but yeah.

File: modules/workflows/OrcaRecoveryWorkflow/orca_recover_workflow.asl.json
Comment (AndrewDorn) [reply modules/workflows/OrcaRecoveryWorkflow/orca_recover_workflow.asl.json:53]: Huh... So the same value was being passed in two different ways, I guess. Alright. Removed.

**Review by bhazuka — APPROVED** [review]
Looks good to me

**Review by rizbi-hassan — APPROVED** [review]

**Pull request merged**
